text/0000-safer-transmute.md

```diff
@@ -0,0 +1,1796 @@
+    }
+};
+```
+Given these two stability promises, we are free to modify the alignment of `Foo` in SemVer-minor changes without running any risk of breaking dependent crates.
+
+
+## Mechanisms of Transmutation
+
+Two traits provide mechanisms for transmutation between types: 
+```rust
+// this trait is implemented automagically by the compiler
+#[lang = "transmute_from"]
+pub unsafe trait TransmuteFrom<Src: ?Sized, Neglect = ()>
+where
+    Neglect: TransmuteOptions,
+{
+    #[inline(always)]
+    fn transmute_from(src: Src) -> Self
+    where
+        Src: Sized,
+        Self: Sized,
+        Neglect: SafeTransmuteOptions,
+    {
+        unsafe { Self::unsafe_transmute_from(src) }
+    }
+
+    #[inline(always)]
+    unsafe fn unsafe_transmute_from(src: Src) -> Self
+    where
+        Src: Sized,
+        Self: Sized,
+        Neglect: TransmuteOptions,
+    {
+        use core::mem::ManuallyDrop;
+
+        #[repr(C)]
+        union Transmute<Src, Dst> {
+            src: ManuallyDrop<Src>,
+            dst: ManuallyDrop<Dst>,
+        }
+
+        unsafe {
+            ManuallyDrop::into_inner(Transmute { src: ManuallyDrop::new(src) }.dst)
+        }
+    }
+}
+
+// implemented in terms of `TransmuteFrom`
+pub unsafe trait TransmuteInto<Dst: ?Sized, Neglect = ()>
+where
+    Neglect: TransmuteOptions,
+{
+    fn transmute_into(self) -> Dst
+    where
+        Self: Sized,
+        Dst: Sized,
+        Neglect: SafeTransmuteOptions;
+
+    unsafe fn unsafe_transmute_into(self) -> Dst
+    where
+        Self: Sized,
+        Dst: Sized,
+        Neglect: TransmuteOptions;
+}
+
+unsafe impl<Src, Dst, Neglect> TransmuteInto<Dst, Neglect> for Src
+where
+    Src: ?Sized,
+    Dst: ?Sized + TransmuteFrom<Src, Neglect>,
+    Neglect: TransmuteOptions,
+{
+    ...
+}
+```
+
+In the above definitions, `Src` represents the source type of the transmutation, `Dst` represents the destination type of the transmutation, and `Neglect` is a parameter that [encodes][options] which static checks the compiler ought to neglect when considering if a transmutation is valid. The default value of `Neglect` is `()`, which reflects that, by default, the compiler does not neglect *any* static checks.
+
+### Neglecting Static Checks
+[options]: #Neglecting-Static-Checks
+
+The default value of the `Neglect` parameter, `()`, statically forbids transmutes that are unsafe, unsound, or unstable. However, you may explicitly opt-out of some static checks:
+
+| Transmute Option    | Compromises | Usable With                                             |
+|---------------------|-------------|---------------------------------------------------------|
+| `NeglectStabilty`   | Stability   | `transmute_{from,into}`, `unsafe_transmute_{from,into}` |
+| `NeglectAlignment`  | Safety      | `unsafe_transmute_{from,into}`                          |
+| `NeglectValidity`   | Soundness   | `unsafe_transmute_{from,into}`                          |
```

Comment:
The names are specific, because they communicate the specific responsibility you assume by neglecting each check. If you `NeglectValidity`, the onus is on you to dynamically ensure `Src` is a bit-valid instance of `Dst` before transmuting. Likewise, if you `NeglectAlignment`, the onus is on you to check alignment (before transmuting in the case of references, and before dereferencing in the case of raw pointers — the "Compromises" reality is perhaps a little more nuanced than the column suggests).

Using specific names also reduces the risk of "overloading" an option with meaning. For instance, a future [`NeglectConstructability`](https://github.com/jswrenn/rfcs/blob/safer-transmute/text/0000-safer-transmute.md#design-of-neglectconstructability) option would *also* compromise safety.

Finally, the boundary between soundness and safety may change over time; things which are soundness properties *could* graduate to safety properties. For instance, while it is [*currently*](https://github.com/rust-lang/rfcs/pull/2981#discussion_r484055005) to UB to create references that are mis-aligned, future versions of Rust plausibly phrase this as a dynamic safety property instead: that is is UB to *de-reference* misaligned referenced. **I don't expect this will happen, in this instance, but it's not unheard of.** I think I recall some recent conversations about whether the bit-validity of `str` could relax from being a soundness issue to a safety issue...

Anyways, I think the "Compromises" column makes the model seem a bit neater than it really is, and should probably be omitted.